theano/tensor/nnet/corr.py

```diff
@@ -0,0 +1,621 @@
+import os
+import logging
+
+import theano
+from theano import Apply
+from theano import gof
+from theano.tensor import as_tensor_variable, TensorType
+from theano.tensor.blas_headers import blas_header_text
+from theano.tensor.blas import ldflags
+
+
+_logger = logging.getLogger(__name__)
+
+
+class BaseCorrMM(gof.Op):
+    """Base class for `CorrMM`, `CorrMM_gradWeights` and
+    `CorrMM_gradInputs`. Cannot be used directly.
+
+    :param border_mode: one of 'valid', 'full', 'half'; additionally, the
+        padding size could be directly specified by an integer or a pair of
+        integers
+    :param subsample: perform subsampling of the output (default: (1, 1))
+    :param pad: *deprecated*, now you should always use border_mode
+
+    """
+    check_broadcast = False
+    __props__ = ('border_mode', 'subsample')
+
+    def __init__(self, border_mode="valid", subsample=(1, 1), pad=(0, 0)):
+        if pad != (0, 0):
+            _logger.warning(
+                'do not use pad for BaseCorrMM; please set padding in '
+                'border_mode parameter, see the docstring for more details')
+            if border_mode != "valid":
+                raise ValueError("border_mode must be 'valid'")
+            border_mode = pad
+        if isinstance(border_mode, int):
+            border_mode = (border_mode, border_mode)
+        if isinstance(border_mode, tuple):
+            pad_h, pad_w = map(int, border_mode)
```

Comment:
We should probably check that the padding passed by the user is >= 0 here, otherwise a user could try passing (-1, -1), expecting the image to be cropped by 1 pixel on each side, but it would actually mean "half" convolution.